File: 15-3sum/15-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
  nums.sort((a, b) => a - b);
  const n = nums.length;
  const ans = [];
  
  for (let i = 0; i < n - 2; i++) {
    if (nums[i] === nums[i - 1]) continue;
    
    let [l, r] = [i + 1, n - 1];
    while (l < r) {
      const sum = nums[i] + nums[l] + nums[r];
      if (!sum) {
        ans.push([nums[i], nums[l], nums[r]]);
        while (nums[l] === nums[++l]);
      }
      
      sum < 0 ? l++ : r--;
    }
  }
  
  return ans;
};

File: 15-3sum/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        ans = []
        
        for i in range(n):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            l, r = i + 1, n -1
            while l < r:
                sum = nums[i] + nums[l] + nums[r]
                if sum == 0:
                    ans.append((nums[i], nums[l], nums[r]))
                    while l < r and nums[l] == nums[l + 1]:
                        l += 1
                    l += 1
                    r -= 1
                elif sum < 0:
                    l += 1
                else:
                    r -= 1
        return ans